homeassistant/components/backup/manager.py

```diff
@@ -453,17 +423,92 @@ async def async_create_backup(
         password: str | None,
         **kwargs: Any,
     ) -> NewBackup:
-        """Initiate generating a backup."""
+        """Initiate generating a backup.
+
+        :param on_progress: A callback that will be called with the progress of the
+            backup.
+        """
         if self.backup_task:
             raise HomeAssistantError("Backup already in progress")
         if not agent_ids:
             raise HomeAssistantError("At least one agent must be selected")
         if any(agent_id not in self.backup_agents for agent_id in agent_ids):
             raise HomeAssistantError("Invalid agent selected")
         backup_name = name or f"Core {HAVERSION}"
+        new_backup, self.backup_task = await self._reader_writer.async_create_backup(
+            addons_included=addons_included,
+            agent_ids=agent_ids,
+            backup_name=backup_name,
+            database_included=database_included,
+            folders_included=folders_included,
+            on_progress=on_progress,
+            password=password,
+        )
+        self.backup_task.add_done_callback(lambda _: setattr(self, "backup_task", None))
+        return new_backup
+
+    async def async_restore_backup(
+        self,
+        backup_id: str,
+        *,
+        agent_id: str,
+        password: str | None,
+        **kwargs: Any,
+    ) -> None:
+        """Initiate restoring a backup.
+
+        :param on_progress: A callback that will be called with the progress of the
+            restore. Home Assistant Core may need to be restarted during the backup
+            restore process, which means the restore process may not be able to report
+            when it's done.
+        """
+
+        if agent_id in self.local_backup_agents:
+            local_agent = self.local_backup_agents[agent_id]
+            if not await local_agent.async_get_backup(backup_id):
+                raise HomeAssistantError(
+                    f"Backup {backup_id} not found in agent {agent_id}"
+                )
+        else:
+            path = self.temp_backup_dir / f"{backup_id}.tar"
+            agent = self.backup_agents[agent_id]
+            if not await agent.async_get_backup(backup_id):
+                raise HomeAssistantError(
+                    f"Backup {backup_id} not found in agent {agent_id}"
+                )
+            await agent.async_download_backup(backup_id, path=path)
+
+        await self._reader_writer.async_restore_backup(
+            backup_id=backup_id,
+            agent_id=agent_id,
+            password=password,
+        )
```

Comment:
I'm not fully sure about the responsibility split and interface between manager and the reader/writer when restoring.
I think it makes sense if the manager checks if the backup is available and downloads it from a remote agent if needed.

In case of the core reader/writer, we just need to know the path to the backup, but that might not work for supervisor.